src/lib/utils/prepareTemplateProps.ts

```diff
@@ -4,59 +4,65 @@ import { Template } from '../../types';
 type TemplatesConfig = object;
 
 type Templates = {
-  [key: string]: Template;
+  [key: string]: Template<any>;
 };
 
-type TemplateProps = {
-  defaultTemplates: Templates;
-  templates: Templates;
+export type PreparedTemplateProps<TTemplates extends Templates> = {
   templatesConfig: TemplatesConfig;
+  templates: TTemplates;
+  useCustomCompileOptions: {
+    [TKey in keyof Partial<TTemplates>]: boolean;
+  };
 };
 
-type PreparedTemplateProps = {
-  templatesConfig: TemplatesConfig;
-  templates: Templates;
-  useCustomCompileOptions: { [key: string]: boolean };
-};
-
-function prepareTemplates(
-  defaultTemplates: Templates = {},
-  templates: Templates = {}
+function prepareTemplates<TTemplates extends Templates>(
+  // can not use = {} here, since the template could have different constraints
+  defaultTemplates?: TTemplates,
```

Comment:
not super happy with this, but refinementList for example doesn't pass defaultTemplates at all